### PR TITLE
switch to dune

### PR DESCRIPTION
**Pull request by redianthus**
Hi,

I just switched the project to dune.

Cheers.

### CODE REVIEW

**Comment by bikallem** [comment]
Will this work with ocaml compiler 4.08 and above. Opam repo seems to constrain this to less than 4.08. However, I don't see such constraint in the opam file itself. 

Perhaps, we can merge the generation of opam into dune-project file itself since we are using dune 1.11. More details here - https://dune.readthedocs.io/en/stable/opam.html?highlight=opam#generating-opam-files

**Comment by redianthus** [comment]
I did a mistake. I changed the `public-name` field from `core_lwt` to `core-lwt` (and also, the opam file name) so now this part should be OK.

For some reasons I'm getting weird errors now:

```sh
dune build
File "lib/core_lwt_stream.ml", line 1:
Error (alert deprecated): on_terminate
 Bind on Lwt_stream.closed.
File "src/core/lwt_stream.mli", line 263, characters 0-100:
  Definition
File "lib/core_lwt_stream.mli", line 166, characters 0-49:
  Expected signature
```

File: lib/dune [hunk 1]
```diff
@@ -0,0 +1,6 @@
+(library
+  (name core_lwt)
+  (public_name core-lwt)
```
Comment (ivg) [inline lib/dune]: I'm not sure that my understanding of the `public_name` stanza is correct, but I think it should `core_lwt`, because this is the name it was known in the [oasis][1] and this is how it was referenced. 

[1]: https://github.com/BinaryAnalysisPlatform/core-lwt/pull/5/files#diff-0b6213b84518a662543be7b7516bc846L13

File: dune-project [hunk 2]
```diff
@@ -0,0 +1,3 @@
+(lang dune 1.11)
+(name core-lwt)
```
Comment (ivg) [inline dune-project]: or `core_lwt`? ;)

**Review by ivg — CHANGES_REQUESTED** [review]

**Comment by ivg** [comment]
> For some reasons I'm getting weird errors now:

It is not an error, but a deprecation alert and dune by default turns all warnings into errors. 
[Here is][1] how to disable it.

[1]: https://dune.readthedocs.io/en/stable/faq.html?highlight=warnings#how-to-make-warnings-non-fatal

**Comment by ivg** [comment]
> I did a mistake. I changed the public-name field from core_lwt to core-lwt (and also, the opam file name) so now this part should be OK.

So the naming story is a little bit confusing, but we should preserve the existing names no matter whether the original choice was right or not. Here is the end result we should strive for:
1) the name of the opam package is `core-lwt`
2) the name of the library is `core_lwt` (with underscore)
3) the name of the `Lwt` module is `Core_lwt.Std.Lwt`

**Comment by ivg** [comment]
And thanks for submitting this. I'm in general in favor of accepting this, once the naming story is fixed. But this is @gitoleg's turf so it is his call. Also, guys, do not forget to package and release this to opam :)

**Comment by bikallem** [comment]
Also please increase ocaml compiler version so that we can use 4.08.1 compiler and above.

**Comment by redianthus** [comment]
OK, I was thinking it was two distinct errors. The deprecation one and then another about some signature mismatch. Actually, it's just the deprecation.

**Comment by redianthus** [comment]
> Also please increase ocaml compiler version so that we can use 4.08.1 compiler and above.

Where ? There's no upper bound on the compiler in the opam file...

**Comment by bikallem** [comment]
opam repo seems to specify upper ocaml compiler constraint.
http://opam.ocaml.org/packages/core-lwt/

**Comment by redianthus** [comment]
OK, but I don't see how this is related to this PR.. It just shouldn't be added when published on opam.

File: lib/dune [hunk 3]
```diff
@@ -0,0 +1,10 @@
+(library
+  (name core_lwt)
+  (public_name core_lwt)
```
Comment (gitoleg) [inline lib/dune:3]: please, update `public_name` here: there should be `core-lwt`. `name` stanza is correct.

File: core_lwt.opam [hunk 4]
```diff
@@ -7,27 +7,16 @@ homepage: "https://github.com/BinaryAnalysisPlatform/core-lwt/"
...
-]
-
-install: [
-  ["ocaml" "setup.ml" "-install"]
-]
-
-remove: [
-  ["ocamlfind" "remove" "core_lwt"]
-]
+build: ["dune" "build" "-p" name "-j" jobs]
```
Comment (gitoleg) [inline core_lwt.opam:9]: and we need to rename this file as well - dune looks for `<package>.opam`. so we need `core-lwt.opam`

**Review by gitoleg — CHANGES_REQUESTED** [review]
There are  just a couple of minor issues and we're good to go.

File: lib/dune (hunk 3 above)
Comment (ivg) [reply lib/dune:3]: No, it should be `core_lwt`! Read the comment above.

**Comment by redianthus** [comment]
I can't have the opam file to be named `core-lwt` and the public_name to be `core_lwt`:

```sh
dune build
File "lib/dune", line 3, characters 2-24:
3 |   (public_name core_lwt)
      ^^^^^^^^^^^^^^^^^^^^^^
Error: The current scope doesn't define package "core_lwt".
The only packages for which you can declare elements to be installed in this
directory are:
- core-lwt (because of core-lwt.opam)
Hint: did you mean core-lwt?
```

So, I did what I think is the correct thing with naming. Let me know if you don't think it's OK.

File: lib/dune (hunk 1 above)
Comment (ivg) [reply lib/dune]: It should be still `core_lwt`, because this is the name of the library as seen by ocamlfind, specified in meta, etc.

**Review by ivg — CHANGES_REQUESTED** [review]
Names are still wrong, sorry ;)

**Comment by redianthus** [comment]
@ivg Did you see my last comment ?

**Comment by ivg** [comment]
@zapashcanon, sure I did :) 
But what is important is that the library name is a part of the interface, you can't just rename `libmath.so` to `libscience.so` and expect people to chew this. Therefore, our task is to keep the library name. And the `public_name` stanza defines the library name, and the name of the generated `cmi` and `cm[x]a` files, which a part of the system definition. 

The dune-opam integration made a very questionable decision to unify library and package name, which will actually break a few dozens of packages in the OCaml universe. Including those that are very important to the OCaml infrastructure. I will reach out the dune guys so we will try to rectify this issue, but what is important is that the dune-opam integration is optional (therefore this limitation doesn't actually propagate towards opam itself), so it should really stop us from merging this PR. 

Next thing to consider, is that `<package>.opam` files are now considered obsolete and will deprecated soon. So we should use the `package` stanza instead of the `<package>.opam` file. See [this topic][1] for more details.

With all that said, we have to admit that in our case the mismatch between the package name and the library name is more of an accident, and probably it would be better to rectify it. Given that the package name is a more volatile instance and the process of package renaming is more well-defined and better supported by the current tools (unlike renaming a library, it is basically impossible to specify a dependency on the same library with different names neither in oasis nor in dune), I would prefer to rename the package name eventually.

To summarize:
1) the library name must be `core_lwt`
2) the local package name, as specified in `dune-project` (although it is optional, we can define it for the ease of use), unfortunately, should be also `core_lwt`
3) the official and public package name will be still `core-lwt` and @gitoleg will release this update and start the renaming process of the package. But this is totally out of the scope of this PR, as you rightfully mentioned above. 

And thanks for starting this work, it actually highlighted an important roadblock towards migration of the whole BAP universe to Dune, which I will discuss with the dune developers. 


[1]: https://dune.readthedocs.io/en/stable/installation.html?highlight=package%20stanza#declaring-a-package

**Comment by ivg** [comment]
Besides, you can keep the `opam` name for the opam file then dune will just ignore this file, but it will be still respected by `opam` itself, so that it could be pinned.

**Comment by redianthus** [comment]
> then dune will just ignore this file,

No. I got:

```sh
dune build
File "lib/dune", line 3, characters 2-24:
3 |   (public_name core_lwt)
      ^^^^^^^^^^^^^^^^^^^^^^
Error: You cannot declare items to be installed without adding a
<package>.opam file at the root of your project.
To declare elements to be installed as part of package "core_lwt", add a
"core_lwt.opam" file at the root of your project.
Root of the project as discovered by dune: .
```

EDIT: so, what I suggest is to change the opam file name to `core_lwt.opam`, so the build works. Then, before releasing a new version, someone can take care of updating reverse dep., there's only one package in opam that still depends on `core-lwt` (I checked using [this small tool](https://git.zapashcanon.fr/zapashcanon/still-dep), so it should be straightforward.

**Comment by ivg** [comment]
Yep, that works for me, @gitoleg  what do you think?

**Comment by gitoleg** [comment]
works for me too

**Comment by redianthus** [comment]
OK, so I think this is ready for merging ?

**Comment by gitoleg** [comment]
@zapashcanon 
sure, as soon as I'll make `travis` tests to work again, we'll merge this PR.

**Comment by redianthus** [comment]
@gitoleg ?

**Comment by ivg** [comment]
I think, we should give it a try. Unfortunately, dune is not versatile enough and forces us to rename the libraries according to their conventions. We can tolerate this for core-lwt,but in general it's not acceptable. So we have to put dune aside. In bap we use omake which is extremely fast (probably even faster than dune), so we won't miss dune.
